python/dgl/distributed/optim/pytorch/sparse_optim.py

```diff
@@ -35,16 +46,194 @@ def __init__(self, params, lr):
+            kvstore = emb.kvstore
+            trainers_per_machine = (
+                num_clients // dgl.distributed.get_num_machines()
+            )
+            emb_state_dict = {}
+            part_policy = (
+                emb.part_policy if emb.part_policy else emb._tensor.part_policy
+            )
+            idx = self._get_local_ids(part_policy)
+            if trainers_per_machine > 1:
+                kv_idx_split = th.remainder(idx, trainers_per_machine).long()
+                client_id = kvstore.client_id
+                local_rank = client_id % trainers_per_machine
+                mask = kv_idx_split == local_rank
+                idx = F.boolean_mask(idx, mask)
+            emb_state_dict.update({"ids": idx})
+            emb_state = {}
+            states = (
+                list(self._state[emb.name])
+                if isinstance(self._state[emb.name], tuple)
+                else [self._state[emb.name]]
+            )
+            emb_state = {state.name: state[idx] for state in states}
+            emb_state_dict.update({"states": emb_state})
+            lcoal_state_dict["emb_states"].update({emb.name: emb_state_dict})
+        lcoal_state_dict["params"].update(self._defaults)
+        return lcoal_state_dict
+
+    def load_local_state_dict(self, local_state_dict):
+        """Load the local state from the input state_dict,
+        updating the optimizer as needed..
+
+        Parameters
+        ----------
+        local_state_dict : dict
+            Optimizer state; should be an object returned
+            from a call to local_state_dict()
+
+        See Also
+        --------
+        local_state_dict
+        """
+        for emb_name, emb_state in local_state_dict["emb_states"].items():
+            idx = emb_state["ids"]
+            states = (
+                list(self._state[emb_name])
+                if isinstance(self._state[emb_name], tuple)
+                else [self._state[emb_name]]
+            )
+            if len(emb_state["states"]) != len(states):
+                raise ValueError(
+                    f"loaded state dict has a different number of states"
+                    f" of embedding {emb_name}"
+                )
+            name_to_index = {
+                state.name: index for index, state in enumerate(states, 0)
+            }
+            for name, state in emb_state["states"].items():
+                if name not in name_to_index:
+                    raise ValueError(
+                        "loaded state dict contains a state {name}"
+                        "that can't be found in the optimizer states"
+                    )
+                state_idx = name_to_index[name]
+                state = state.to(states[state_idx].dtype)
+                state = state.to(th.device("cpu"))
+                states[state_idx][idx] = state
+        self._defaults.update(local_state_dict["params"])
+        self.__dict__.update(local_state_dict["params"])
```

Comment:
`_defaults` and `__dict__` (i.e. the attributes of the class) are the same?  Then why do we need `_defaults`?

Reply:
__dict__ contains everything, including items we want and we don't want , but we can't tell them. 
It also follow the PyTorch style to add a _defaults.

Reply:
I didn't see where PyTorch has this.  Could you give a reference?

Reply:
See here[Optimizer base class]( https://github.com/pytorch/pytorch/blob/b136f3f310aa01a8b3c1e63dc0bfda8fd2234b06/torch/optim/optimizer.py) and [Adagrad](https://github.com/pytorch/pytorch/blob/b136f3f310aa01a8b3c1e63dc0bfda8fd2234b06/torch/optim/adagrad.py)